File: backend/src/api_computations.c
/*
 * api_computations.c — API-facing computation wrappers
 *
 * Composes Phase 2a SOLVE_WIDGET primitives for specific user situations.
 * These functions are called by the webserver HTTP handlers.
 */
#include <math.h>
#include <stdio.h>

#include "api_computations.h"
#include "widget_solver.h"
#include "dice_mechanics.h"
#include "game_mechanics.h"

/* Compute best reroll mask for a specific game situation.
 * Builds the reroll-level arrays then picks the best mask for the given dice. */
void ComputeBestRerollStrategy(const YatzyContext *ctx,
                               int upper_score,
                               int scored_categories,
                               const int dice[5],
                               int rerolls_remaining,
                               int *best_mask,
                               double *best_ev) {
    /* Group 6: E(S, r, 0) for all r */
    double E_ds_0[252];
    YatzyState state = {upper_score, scored_categories};
    for (int ds_i = 0; ds_i < 252; ds_i++) {
        E_ds_0[ds_i] = ComputeBestScoringValueForDiceSet(ctx, &state, ctx->all_dice_sets[ds_i]);
    }

    if (rerolls_remaining == 1) {
        *best_mask = ChooseBestRerollMask(ctx, E_ds_0, dice, best_ev);
        return;
    }

    double E_ds_1[252];
    int dummy_mask[252];
    ComputeExpectedValuesForNRerolls(ctx, E_ds_0, E_ds_1, dummy_mask);
    *best_mask = ChooseBestRerollMask(ctx, E_ds_1, dice, best_ev);
}

/* Choose best category when no rerolls remain (non-zero score only). */
int ChooseBestCategoryNoRerolls(const YatzyContext *ctx,
                                int upper_score,
                                int scored_categories,
                                const int dice[5],
                                double *best_ev) {
    int ds_index = FindDiceSetIndex(ctx, dice);
    double best_val = -INFINITY;
    int best_category = -1;

    for (int c = 0; c < CATEGORY_COUNT; c++) {
        if (IS_CATEGORY_SCORED(scored_categories, c)) {
            continue;
        }

        int scr = ctx->precomputed_scores[ds_index][c];

        if (scr == 0) {
            continue;
        }

        int new_up = UpdateUpperScore(upper_score, c, scr);
        int new_scored = scored_categories | (1 << c);
        double val = scr + GetStateValue(ctx, new_up, new_scored);

        if (val > best_val) {
            best_val = val;
            best_category = c;
        }
    }

    *best_ev = best_val;
    return best_category;
}

/* Evaluate the EV of a user-chosen reroll mask. */
double EvaluateChosenRerollMask(const YatzyContext *ctx,
                                int upper_score,
                                int scored_categories,
                                const int dice[5],
                                int chosen_mask,
                                int rerolls_remaining) {
    /* Group 6: E(S, r, 0) for all r */
    double E_ds_0[252];
    YatzyState state = {upper_score, scored_categories};
    for (int ds_i = 0; ds_i < 252; ds_i++) {
        E_ds_0[ds_i] = ComputeBestScoringValueForDiceSet(ctx, &state, ctx->all_dice_sets[ds_i]);
    }

    double E_ds_1[252];
    int dummy_mask[252];
    if (rerolls_remaining == 2) {
        ComputeExpectedValuesForNRerolls(ctx, E_ds_0, E_ds_1, dummy_mask);
    }

    const double *E_ds_for_masks = (rerolls_remaining == 1) ? E_ds_0 : E_ds_1;

    int sorted_dice[5];
    for (int i = 0; i < 5; i++) sorted_dice[i] = dice[i];
    SortDiceSet(sorted_dice);
    int ds_index = FindDiceSetIndex(ctx, sorted_dice);

    double ev = ComputeExpectedValueForRerollMask(ctx, ds_index, E_ds_for_masks, chosen_mask);
    return ev;
}

/* Evaluate the EV of a user-chosen category assignment. */
double EvaluateChosenCategory(const YatzyContext *ctx,
                              int upper_score,
                              int scored_categories,
                              const int dice[5],
                              int chosen_category) {
    if (IS_CATEGORY_SCORED(scored_categories, chosen_category)) {
        return -INFINITY;
    }
    int ds_index = FindDiceSetIndex(ctx, dice);
    int score = ctx->precomputed_scores[ds_index][chosen_category];

    if (score == 0) return -INFINITY;
    int new_up = UpdateUpperScore(upper_score, chosen_category, score);
    int new_scored = scored_categories | (1 << chosen_category);
    double future_val = GetStateValue(ctx, new_up, new_scored);
    return score + future_val;
}

/* Build outcome distribution for a specific reroll mask.
 * Pairs each possible outcome ds2 with its EV and transition probability. */
void ComputeDistributionForRerollMask(const YatzyContext *ctx,
                                      int ds_index,
                                      const double E_ds_for_masks[252],
                                      int mask,
                                      EVProbabilityPair out_distribution[252]) {
    /* Initialize all entries with zero probability */
    for (int ds2_i = 0; ds2_i < 252; ds2_i++) {
        out_distribution[ds2_i].ev = E_ds_for_masks[ds2_i];
        out_distribution[ds2_i].probability = 0.0;
        out_distribution[ds2_i].ds2_index = ds2_i;
    }

    /* Fill non-zero probabilities from sparse table */
    const SparseTransitionTable *sp = &ctx->sparse_transitions;
    int row = ds_index * 32 + mask;
    int start = sp->row_offsets[row];
    int end = sp->row_offsets[row + 1];
    for (int k = start; k < end; k++) {
        out_distribution[sp->col_indices[k]].probability = sp->values[k];
    }
}

/* Compute E(S, r, n) for all r, for a given number of rerolls.
 * Copies result into caller-provided buffer (thread-safe). */
void ComputeExpectedValues(const YatzyContext *ctx, int upper_score,
                           int scored_categories, int rerolls,
                           double out_E_ds[252]) {
    double E_ds[3][252];

    /* Level 0: E(S, r, 0) = max_c [s(S,r,c) + E(n(S,r,c))] */
    #pragma omp parallel for
    for (int ds_i = 0; ds_i < 252; ds_i++) {
        double best_val = -INFINITY;
        for (int c = 0; c < CATEGORY_COUNT; c++) {
            if (!IS_CATEGORY_SCORED(scored_categories, c)) {
                int scr = ctx->precomputed_scores[ds_i][c];
                int new_up = UpdateUpperScore(upper_score, c, scr);
                int new_scored = scored_categories | (1 << c);
                double val = scr + GetStateValue(ctx, new_up, new_scored);
                if (val > best_val) best_val = val;
            }
        }
        E_ds[0][ds_i] = best_val;
    }

    /* Levels 1..rerolls: E(S, r, n) = max_{mask} Σ P(r'→r'') · E_ds[n-1][r''] */
    const double * restrict vals = ctx->sparse_transitions.values;
    const int * restrict cols = ctx->sparse_transitions.col_indices;
    const int *offsets = ctx->sparse_transitions.row_offsets;

    for (int n = 1; n <= rerolls; n++) {
        #pragma omp parallel for
        for (int ds_i = 0; ds_i < 252; ds_i++) {
            double best_val = E_ds[n - 1][ds_i]; /* mask=0: keep all */
            int row_base = ds_i * 32;
            for (int mask = 1; mask < 32; mask++) {
                int start = offsets[row_base + mask];
                int end = offsets[row_base + mask + 1];
                double ev = 0.0;
                for (int k = start; k < end; k++) {
                    ev += vals[k] * E_ds[n - 1][cols[k]];
                }
                if (ev > best_val) best_val = ev;
            }
            E_ds[n][ds_i] = best_val;
        }
    }

    /* Copy result to caller's buffer */
    for (int i = 0; i < 252; i++) {
        out_E_ds[i] = E_ds[rerolls][i];
    }
}

/* Weighted sum of EV over a probability distribution. */
double ComputeEVFromDistribution(const EVProbabilityPair distribution[], int size) {
    double ev = 0.0;
    for (int i = 0; i < size; i++) {
        ev += distribution[i].ev * distribution[i].probability;
    }
    return ev;
}

File: backend/src/api_computations.rs
//! API-facing computation wrappers.
//!
//! The main entry point is [`compute_roll_response`], which returns everything
//! the frontend needs for a single roll: mask EVs, category scores/EVs, and optimal
//! actions. One backend call per roll, zero for dice toggling or category selection.
//!
//! All internal computation uses f32; return types are f64 for JSON compatibility.

use crate::constants::*;
use crate::dice_mechanics::{find_dice_set_index, sort_dice_set};
use crate::game_mechanics::update_upper_score;
use crate::types::YatzyContext;
use crate::widget_solver::*;

/// Per-category info returned in the fat response.
#[derive(Clone)]
pub struct CategoryInfo {
    pub id: usize,
    pub name: &'static str,
    pub score: i32,
    pub available: bool,
    pub ev_if_scored: f64,
}

/// Everything the frontend needs until the next roll.
pub struct RollResponse {
    /// EV for each of the 32 reroll masks (None when rerolls_remaining == 0).
    pub mask_evs: Option<[f64; 32]>,
    /// Best reroll mask (None when rerolls_remaining == 0).
    pub optimal_mask: Option<i32>,
    /// EV of the optimal mask (None when rerolls_remaining == 0).
    pub optimal_mask_ev: Option<f64>,
    /// All 15 categories with scores and EVs.
    pub categories: [CategoryInfo; CATEGORY_COUNT],
    /// Best category to score (argmax of ev_if_scored among available).
    pub optimal_category: i32,
    /// EV of the optimal category.
    pub optimal_category_ev: f64,
    /// Overall state EV (for reference).
    pub state_ev: f64,
}

/// Compute everything the frontend needs for a single roll.
///
/// 1. Group 6: E(S, r, 0) for all 252 dice sets + per-category info for the given dice
/// 2. If rerolls > 0: propagate through Groups 5/3 to get e_ds at the right reroll level
/// 3. Compute all 32 mask EVs for the specific dice set
/// 4. Find optimal mask and optimal category
pub fn compute_roll_response(
    ctx: &YatzyContext,
    upper_score: i32,
    scored_categories: i32,
    dice: &[i32; 5],
    rerolls_remaining: i32,
) -> RollResponse {
    let mut sorted_dice = *dice;
    sort_dice_set(&mut sorted_dice);
    let ds_index = find_dice_set_index(ctx, &sorted_dice);
    let sv = ctx.state_values.as_slice();

    // --- Build categories info ---
    let categories: [CategoryInfo; CATEGORY_COUNT] = std::array::from_fn(|c| {
        let scr = ctx.precomputed_scores[ds_index][c];
        let available = !is_category_scored(scored_categories, c);
        let ev_if_scored = if available {
            let new_up = update_upper_score(upper_score, c, scr);
            let new_scored = scored_categories | (1 << c);
            (scr as f32 + sv[state_index(new_up as usize, new_scored as usize)]) as f64
        } else {
            f64::NEG_INFINITY
        };
        CategoryInfo {
            id: c,
            name: CATEGORY_NAMES[c],
            score: scr,
            available,
            ev_if_scored,
        }
    });

    // Find optimal category (argmax ev_if_scored among available with score > 0)
    let mut optimal_category = -1i32;
    let mut optimal_category_ev = f64::NEG_INFINITY;
    for cat in &categories {
        if cat.available && cat.score > 0 && cat.ev_if_scored > optimal_category_ev {
            optimal_category_ev = cat.ev_if_scored;
            optimal_category = cat.id as i32;
        }
    }

    // --- Build Group 6: E(S, r, 0) for all 252 dice sets ---
    let mut e_ds_0 = [0.0f32; 252];
    for ds_i in 0..252 {
        let mut best_val = f32::NEG_INFINITY;
        for c in 0..CATEGORY_COUNT {
            if !is_category_scored(scored_categories, c) {
                let scr = ctx.precomputed_scores[ds_i][c];
                let new_up = update_upper_score(upper_score, c, scr);
                let new_scored = scored_categories | (1 << c);
                let val = scr as f32 + sv[state_index(new_up as usize, new_scored as usize)];
                if val > best_val {
                    best_val = val;
                }
            }
        }
        e_ds_0[ds_i] = best_val;
    }

    // --- If rerolls remaining, compute mask EVs ---
    if rerolls_remaining > 0 {
        // Build reroll levels
        let e_ds_for_masks = if rerolls_remaining == 1 {
            e_ds_0
        } else {
            // rerolls == 2: need one level of propagation
            let mut e_ds_1 = [0.0f32; 252];
            let mut dummy_mask = [0i32; 252];
            compute_expected_values_for_n_rerolls(ctx, &e_ds_0, &mut e_ds_1, &mut dummy_mask);
            e_ds_1
        };

        // Compute all 32 mask EVs
        let mut mask_evs = [0.0f64; 32];
        for mask in 0..32 {
            mask_evs[mask] =
                compute_expected_value_for_reroll_mask(ctx, ds_index, &e_ds_for_masks, mask as i32);
        }

        // Find optimal mask
        let mut best_mask = 0i32;
        let mut best_mask_ev = mask_evs[0];
        for mask in 1..32 {
            if mask_evs[mask] > best_mask_ev {
                best_mask_ev = mask_evs[mask];
                best_mask = mask as i32;
            }
        }

        // Compare optimal mask vs optimal category to set overall state_ev
        let effective_state_ev = best_mask_ev.max(optimal_category_ev);

        RollResponse {
            mask_evs: Some(mask_evs),
            optimal_mask: Some(best_mask),
            optimal_mask_ev: Some(best_mask_ev),
            categories,
            optimal_category,
            optimal_category_ev,
            state_ev: effective_state_ev,
        }
    } else {
        // No rerolls: only categories matter
        RollResponse {
            mask_evs: None,
            optimal_mask: None,
            optimal_mask_ev: None,
            categories,
            optimal_category,
            optimal_category_ev,
            state_ev: optimal_category_ev,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::phase0_tables;

    fn make_ctx() -> Box<YatzyContext> {
        let mut ctx = YatzyContext::new_boxed();
        phase0_tables::precompute_lookup_tables(&mut ctx);
        ctx
    }

    #[test]
    fn test_fat_response_with_rerolls() {
        let ctx = make_ctx();
        let dice = [1, 2, 3, 4, 5];
        let resp = compute_roll_response(&ctx, 0, 0, &dice, 2);

        // Should have mask EVs since rerolls > 0
        assert!(resp.mask_evs.is_some());
        assert!(resp.optimal_mask.is_some());
        assert!(resp.optimal_mask_ev.is_some());

        // All 15 categories should be available
        for cat in &resp.categories {
            assert!(cat.available);
        }

        // mask_evs[0] (keep all) should equal the best category EV
        let mask_evs = resp.mask_evs.unwrap();
        assert!(mask_evs[0].is_finite());

        // Optimal mask EV should be >= keep-all EV
        assert!(resp.optimal_mask_ev.unwrap() >= mask_evs[0] - 1e-9);
    }

    #[test]
    fn test_fat_response_no_rerolls() {
        let ctx = make_ctx();
        let dice = [6, 6, 6, 6, 6];
        let resp = compute_roll_response(&ctx, 0, 0, &dice, 0);

        // Should NOT have mask EVs
        assert!(resp.mask_evs.is_none());
        assert!(resp.optimal_mask.is_none());
        assert!(resp.optimal_mask_ev.is_none());

        // All categories available, Yatzy should score 50
        let yatzy = &resp.categories[CATEGORY_YATZY];
        assert!(yatzy.available);
        assert_eq!(yatzy.score, 50);
    }

    #[test]
    fn test_fat_response_scored_categories() {
        let ctx = make_ctx();
        let scored = (1 << CATEGORY_ONES) | (1 << CATEGORY_TWOS);
        let dice = [3, 3, 4, 5, 6];
        let resp = compute_roll_response(&ctx, 5, scored, &dice, 1);

        // Ones and Twos should be unavailable
        assert!(!resp.categories[CATEGORY_ONES].available);
        assert!(!resp.categories[CATEGORY_TWOS].available);
        assert!(resp.categories[CATEGORY_ONES].ev_if_scored.is_infinite());

        // Remaining categories should be available
        for c in 2..CATEGORY_COUNT {
            assert!(resp.categories[c].available);
        }
    }

    #[test]
    fn test_fat_response_category_scores() {
        let ctx = make_ctx();
        let dice = [1, 2, 3, 4, 5];
        let resp = compute_roll_response(&ctx, 0, 0, &dice, 0);

        // Verify known scores
        assert_eq!(resp.categories[CATEGORY_ONES].score, 1);
        assert_eq!(resp.categories[CATEGORY_TWOS].score, 2);
        assert_eq!(resp.categories[CATEGORY_SMALL_STRAIGHT].score, 15);
        assert_eq!(resp.categories[CATEGORY_LARGE_STRAIGHT].score, 0);
        assert_eq!(resp.categories[CATEGORY_YATZY].score, 0);
        assert_eq!(resp.categories[CATEGORY_CHANCE].score, 15);
    }

    #[test]
    fn test_fat_response_optimal_category_yatzy() {
        let ctx = make_ctx();
        // Only Yatzy + Chance left, roll all sixes
        let all_scored = (1 << CATEGORY_COUNT) - 1;
        let scored = all_scored ^ (1 << CATEGORY_YATZY) ^ (1 << CATEGORY_CHANCE);
        let dice = [6, 6, 6, 6, 6];
        let resp = compute_roll_response(&ctx, 0, scored, &dice, 0);

        assert_eq!(resp.optimal_category, CATEGORY_YATZY as i32);
    }
}
